FS: Socket setup

navjot-chahal opened the pull request:
### What this PR does:
- Set up socket.io for FE and BE.
- Added auth to the sockets handshake.
- Set up "online users" functionality.
- Designed a module with userLogger class for storing currently online user data.
- Console logs are left as placeholders for future code and must be removed later on.

### Screenshots / Videos:
- Browser console log after testing logging with three separate users in different windows.

<img width="260" alt="Socket Response" src="https://user-images.githubusercontent.com/60498899/139047458-9578a047-bbdd-4657-b625-deb793b4c407.png">

- Node.js console log after testing logging with three separate users in different windows.

![Node js log](https://user-images.githubusercontent.com/60498899/139047511-504909fd-4b18-44c8-b55d-9d16a4074eb6.png)

### Any information needed to test this feature:
- This is mostly BE and IN setup and most of the code feedback (console.log statements) have been removed.

closes #30

navjot-chahal committed:
FS: Socket setup

navjot-chahal committed:
Merge branch 'develop' into feature_add_FS-socketSetup

sundayezeilo commented on client/src/context/useSocketContext.tsx:
```diff
@@ -4,26 +4,46 @@ import { io, Socket } from 'socket.io-client';
+    newSocket.on('newUserOnline', (newUser) => {
+      // console.log('New user is online now!');
+      // console.log(newUser);
+    });
```
Noted you will remove all these comments

sundayezeilo commented on server/utils/onlineUserLog.js:
```diff
@@ -0,0 +1,47 @@
+
+class onlineUserLog {
+  constructor() {
+    this.log = new Array();
```
Using an array to log online users quite works. But it would take a shorter time to find a user in an object than in an array, especially when the number increases largely. So, maybe you can consider using an object instead.

navjot-chahal replied:
I was scratching my head when you mentioned objects will take constant time for lookup. As per my understanding of data structures in C++, only Hashtables have O(1) search time. Then upon searching i realized objects use hashtable for underlying data structure😂. Sure, I will implement it now. Nice detail to learn, thanks.

sundayezeilo replied:
That's cool, @navjot-chahal. All `key/value` pair data structures have O(1) access time.

sundayezeilo commented on server/utils/onlineUserLog.js:
```diff
@@ -0,0 +1,47 @@
+      }
+    }
+    return false;
+  }
```
This would rather take constant time to do with a JavaScript object.

sundayezeilo reviewed (commented):
Nice one here :+1: There are just a few comments to take a look at.

navjot-chahal committed:
Hotfix - 1

navjot-chahal committed:
Merge branch 'feature_add_FS-socketSetup' of https://github.com/hatchways/team-scallops into feature_add_FS-socketSetup

navjot-chahal commented:
@ezeilo-su, 

- I have changed the underlying user logging data structure to Object-based (Hashtable) instead of array-based for faster read times. See `Hotfix - 1` for your reference.
- I added a `userDisconnected` `emit` event for user disconnect as well.

rajivtitus commented on server/app.js:
```diff
@@ -28,11 +30,29 @@ const server = http.createServer(app);
+      const newUser = onlineUserLog.addUser(socket.user, socket.id);
+      if (!newUser) return;
+      socket.join("online");
+      socket.to("online").emit("newUserOnline", newUser);
```
Why don't we use `socket.broadcast.emit()` as it behaves similar to io.sockets.emit, but instead of emitting to all connected sockets it will emit to all connected socket except the one it is being called on - seems more appropriate 😄

navjot-chahal replied:
@rajivtitus, I am not fully sure about your concern here. 
Did you mean we should replace `socket.to("online").emit("newUserOnline", newUser);` by `socket.broadcast.to("online").emit("newUserOnline", newUser);`. If yes, then there is no need actually.

`socket.to(...)` **presumes** `socket.broadcast.to(...)` before emitting, thus the sender themself does not receive the emitted event. [Here ](https://socket.io/docs/v3/emit-cheatsheet/)is the reference.

Also, let me know if you mean otherwise. Nonetheless, I am removing `online` room altogether in my next PR as a connected socket is always online.

rajivtitus replied:
@navjot-chahal Thanks for the clarification and reference 👍🏽

rajivtitus commented on server/app.js:
```diff
@@ -28,11 +30,29 @@ const server = http.createServer(app);
+    const deletedUser = onlineUserLog.removeUser(socket.user);
+
+    if (!deletedUser) return;
+    socket.to("online").emit("userDisconnected", deletedUser);
```
Same as comment below

rajivtitus reviewed (approved):
Good work here! Just a comment for you to have a look and address. Thanks!
